app/code/community/Bolt/Boltpay/controllers/ApiController.php

```diff
@@ -80,9 +80,29 @@ public function hookAction()
             $orderId = $transaction->order->cart->display_id;
             $quoteId = $transaction->order->cart->order_reference;
 
-            $order = Mage::getModel('sales/order')->loadByIncrementId($orderId);
-
-            if (!empty($order->getData())) {
+            /***************************************************************************
+             * Handle the racing condition of frontend and Bolt Server order creation
+             * The key idea is that we want the front-side order creation to succeed
+             * always before hooks to give the end customer the best user experience.
+             * Therefore, we sleep execution for a short time over a limited number
+             * attempts if the hook cannot find the order.
+             *
+             * We also ignore Bolt aborts so that if a reasonable amount of time has
+             * passed with no results, it is safe to assume the front-end order creation
+             * failed to deliver on user experience and we should go ahead and treat
+             * it as an orphaned transaction and create the order.
+             ***************************************************************************/
+            ignore_user_abort(true);
+            set_time_limit( 60 );
+            $remaining_attempts = 6;
+            do {
+                /* @var Mage_Sales_Model_Order $order */
+                $order = Mage::getModel('sales/order')->loadByIncrementId($orderId);
+                sleep(5 );
```

Comment:
Would it be simpler to just increase the time that Bolt waits before the initial attempt to call the hook api? I seem to recall that Kazuki can change this on the Bolt side. Just worried that if we do it here we disable their ability to do it on the Bolt side.

Reply:
Yes let's increase delay for calling hook from bolt. It's currently 10sec. What do you suggest? 30sec? 60sec?

Sleeping is not a good idea because 
1) process keeps holding resources (eg., TCP connections) and 2) caller (in this case Bolt server) may timeout.

Reply:
Most importantly, I will first say that a Bolt side implementation for increased delay would definitely be ideal. Historically, It started at 0 seconds, moved to 5 seconds, and was quickly switched to 10 seconds.  Still, none of these are enough, and I recall Kazuki mentioning to me that he has seen times well above that.  In the code above, We give it 30 seconds and I believe 30 seconds is more than enough.  If it is taking that long, there are other merchant-side problems to consider.

Now, if in reality, it is difficult to get this change switched on Bolt-side for whatever reason, then it appears useful to explain some things about the code above.

@johnmccombs1 In no way can this disable the ability on Bolt-side to create a delay.  In the average case scenario, when Bolt arrives, there is no delay.  I would guess in the high 90 percents and maybe much more.  In fact, if caching of shipping and tax is done with foresight, and used here, then I think that this would reach near 100%.  The next most common scenario is Bolt time plus 5 seconds.

@daisy1754  Bolt-side solution is certainly desirable across all plugins, and if that code is set to 30 seconds, then this fallback will likely never be triggered.  But, if for whatever reason it is triggered:

1.)  PHP `sleep` does not take up any CPU resources.  The only resources would be the memory footprint used by Magento and we would hope that this is negligable in a mature software framework, and perhaps, expect it.

2.) The code assumes that Bolt will drop the connection before it finishes.
` ignore_user_abort(true);` handles this.  It is coupled with a maximum allowed execution time of 60 seconds by `set_time_limit( 60 );`  TCP connections end as soon as Bolt ends them, which I also believe max time is 10 seconds.

Finally, in practice, this code should be triggered very rarely, but when used, it makes the current Bolt-side 10 second delay practical (or even better a 5 second delay) and can reduce the auth confirmation time from the proposed 30 delay + plus magento processing time to something closer to de facto real time.  
Most importantly, I will first say that a Bolt side implementation for increased delay would definitely be ideal. Historically, It started at 0 seconds, moved to 5 seconds, and was quickly switched to 10 seconds.  Still, none of these are enough, and I recall Kazuki mentioning to me that he has seen times well above that.  In the code above, We give it 30 seconds and I believe 30 seconds is more than enough.  If it is taking that long, there are other merchant-side problems to consider.

Now, if in reality, it is difficult to get this change switched on Bolt-side for whatever reason, then it appears useful to explain some things about the code above.

@johnmccombs1 In no way can this disable the ability on Bolt-side to create a delay.  In the average case scenario, when Bolt arrives, there is no delay.  I would guess in the high 90 percents and maybe much more.  In fact, if caching of shipping and tax is done with foresight, and used here, then I think that this would reach near 100%.  The next most common scenario is Bolt time plus 5 seconds.

@daisy1754  Bolt-side solution is certainly desirable across all plugins, and if that code is set to 30 seconds, then this fallback will likely never be triggered.  But, if for whatever reason it is triggered:

1.)  PHP `sleep` does not take up any CPU resources.  The only resources would be the memory footprint used by Magento and we would hope that this is negligable in a mature software framework, and perhaps, expect it.

2.) The code assumes that Bolt will drop the connection before it finishes.
` ignore_user_abort(true);` handles this.  It is coupled with a maximum allowed execution time of 60 seconds by `set_time_limit( 60 );`  TCP connections end as soon as Bolt ends them, which I also believe max time is 10 seconds.

Finally, in practice, this code should be triggered very rarely, but when used, it makes the current Bolt-side 10 second delay practical (or even better a 5 second delay) and can reduce the auth confirmation time from the proposed 30 delay + plus magento processing time to something closer to de facto real time.

Reply:
I meant HTTP connection resource. Even if machine is idle, it'll keep HTTP connection for 5+ seconds which is not ideal.
Also IMO this makes code more complex to deal with theoretical scenario.


Also I'm not sure what we're optimizing here for. If for some reason hook creates order first but frontend makes AJAX calls magento for order creation later, shouldn't AJAX call returns success response and frontend should show success message?

Reply:
@daisy1754 https://github.com/BoltApp/bolt-magento1/pull/54